### PR TITLE
ReScript Project spec

### PR DESCRIPTION
**Pull request by IwanKaramazow**
Formalize the heuristics to detect compiler executables for editor tooling

### CODE REVIEW

**Comment by IwanKaramazow** [comment]
@bobzhang could you give this a review? Editor tooling needs to be able to get this correct for monorepos and normal projects.

File: PROJECT_HEURISTICS.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+## Heuristics
+
+1. `.compiler.log` (diagnostics)
+- Starting from the current open file, walk the filesystem upwards, until `bsconfig.json` is located
+- `.compiler.log` is located at `path.join(foundDirectory, "lib", "bs", ".compiler.log")`
+
+2. platform specific `bsc.exe` (formatter)
+- Starting from the current open file, walk the filesystem upwards, until `./node_modules/bs-platform` is located
+- `bsc.exe` is located at `path.join(foundDirectory, "./node_modules/bs-platform", platform, "bsc.exe")`
+
```
Comment (bobzhang) [inline PROJECT_HEURISTICS.md:15]: This is not exact since if you are not in the root project but in a dependency, this might be wrong

Comment (bobzhang) [reply PROJECT_HEURISTICS.md:15]: I have a vague memory that there is something called `path.resolve`

File: PROJECT_HEURISTICS.md [hunk 2]
```diff
@@ -0,0 +1,261 @@
+# Project heuristics
+
+Vscode and vim need to be able to detect the correct location of the `.compiler.log`, `bsc.exe` and `bsb` for diagnostics and formatting.
+Different project layouts result into different locations. The aim of this document is to formalize the different cases and specify where the editor tooling should search for `.compiler.log`, `bsc.exe` and `bsb`
+
+## Heuristics
+
+1. `.compiler.log` (diagnostics)
+- Starting from the current open file, walk the filesystem upwards, until `bsconfig.json` is located
+- `.compiler.log` is located at `path.join(foundDirectory, "lib", "bs", ".compiler.log")`
```
Comment (bobzhang) [inline PROJECT_HEURISTICS.md:10]: this is correct

File: PROJECT_HEURISTICS.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+Typical workflow:
+1. run `yarn` in the root
+2. build folder1: `cd folder1 && yarn bsb`
+
+When `/folder1/src/App.res` is openend in the editor:
+1. `bsc.exe` can be found in `/root/node_modules/bs-platform/{platform}/bsc.exe`
+2. `.compiler.log` can be found in `/root/folder1/lib/bs/.compiler.log`
+3. `bsb` to get a build for this project and get the freshest data: `/root/folder1/node_modules/.bin/bsb`. Note that there's also a one in `/root/node_modules/.bin/bsb`. The bsb in the subfolder is a symlink to the one in the root's node_modules. Bsb needs a `bsconfig.json` in the current working directory, so the cwd for bsb needs to be `/root/folder1`
+
```
Comment (bobzhang) [inline PROJECT_HEURISTICS.md:82]: seems correct

File: PROJECT_HEURISTICS.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+      bsb
+  package.json
+  yarn.lock
+```
+
+Typical workflow:
+1. run `yarn` in the root
+2. build folder1: `cd folder1 && yarn bsb`
+3. build folder2: `cd folder2 && yarn bsb`
+
```
Comment (bobzhang) [inline PROJECT_HEURISTICS.md:125]: this is incorrect. user should only run bsb in the main repo.
The project is organised with the assumption that there's only main entry point, user should never run `bsb` in different directories.

File: PROJECT_HEURISTICS.md [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+When `/root/folder2/src/Folder2App.res` is openend in the editor:
+1. `bsc.exe` can be found in `/root/node_modules/bs-platform/{platform}/bsc.exe`
+2. `.compiler.log` can be found in `/root/folder2/lib/bs/.compiler.log`
+3. `bsb` to get a build for this project and get the freshest data: `/root/folder2/node_modules/.bin/bsb`. Note that there's also a one in `/root/node_modules/.bin/bsb`. The bsb in the subfolder is a symlink to the one in the root's node_modules. Bsb needs a `bsconfig.json` in the current working directory, so the cwd for bsb needs to be `/root/folder2`
+
+### Yarn workspace with multiple ReScript projects in subfolders where one is on a different bs-platform version
+
+Monorepo where `folder1` and `folder2` are on `bs-platform` 8.4.2 and `folder3` on 8.3.3
+
```
Comment (bobzhang) [inline PROJECT_HEURISTICS.md:140]: This should not happen, since the main entry point decides the compiler version

**Review by bobzhang — CHANGES_REQUESTED** [review]
we can schedule a meeting if you find it more efficient

**Comment by IwanKaramazow** [comment]
@bobzhang if you have some time, would it be possible to schedule a short meeting to go over my questions and your feedback? It might be faster.

**Comment by chenglou** [comment]
Superseded by #59 and 5caa099c7d3133fa8e444dd1162388a4c34f6b7b. Closing for now til we reopen this.

**Pull request closed without merging**